### DISCUSSION TITLE
Possible to change naming and default value forms?

### DISCUSSION THREAD

**Discussion by JohanPetersson** [dotnet/templating#4500] [Template authoring] [answered]
Hi, is it possible to change what transformation is used when replacing `sourceName` with `name`? I have a file with the same name as `sourceName` and if `name` contains a whitespace it gets replace with a hyphen '-'.  I want to reference this filename in a config file, but in config files the `sourceName` gets replace with an underscore '_'. See https://github.com/dotnet/templating/wiki/Naming-and-default-value-forms.  Or can I achieve this in another way?

Edit: After posting this, I realize that I could probably use a derived symbol from `name` and rename the file by replacing '-' with '_', https://github.com/dotnet/templating/wiki/Reference-for-template.json#derived-symbol.

**Comment by vlada-shubina** [comment, marked as answer]
@JohanPetersson which `sourceName` do you use in your template configuration? 
By using unambiguous source name (as `Template.1` from https://github.com/dotnet/templating/wiki/Naming-and-default-value-forms) you can control applied value forms.

In case you need safe namespace form to be applied in template content, use `Template._1` in the content.
In case you need name as stated by user (identity form) in template content, use `Template.1` in the content.

**Comment by JohanPetersson** [comment]
I'm not sure I fully understand those tables :( My `sourceName` is e.g. Test.Test. Ended up with this solution:

```json
"symbols": {
  "safeFilename": {
    "type": "derived",
    "valueSource": "name",
    "valueTransform": "replaceHyphensWithUnderscore",
    "fileRename": "Test.Test",
    "description": "Renames files with underscores instead of hyphens so it matches the names in e.g. appSettings.json"
  },
},
"forms": {
  "replaceHyphensWithUnderscore": {
    "identifier": "replace",
    "pattern": "(-)+",
    "replacement": "_"
  }
}
```

Unfortunately, this touches all files with the matching `sourceName`. Ideally I would just want to touch one filetype (file extension). But if I add a file extension in `fileRename` it all breaks.

**Comment by JohanPetersson** [comment]
I have a file called e.g. Test.Test.mdf and a connection string in appSettings.json:

```json
"ConnectionStrings": {
  "db": "Data Source=(LocalDb)\\MSSQLLocalDB;AttachDbFilename=|DataDirectory|\\Test.Test.mdf;Initial Catalog=Test.Test;Integrated Security=True;Connect Timeout=30"
}
```

Given the `name` 'new-name' I get new-name.mdf and:
```json
"ConnectionStrings": {
  "db": "Data Source=(LocalDb)\\MSSQLLocalDB;AttachDbFilename=|DataDirectory|\\new_name.mdf;Initial Catalog=new_name;Integrated Security=True;Connect Timeout=30"
}
```

With my fix above the mdf file gets named new_name.mdf so the connection string is correct.

Do you mean I could name my file Test._Test.mdf instead and use the same format (Test._Test) in the connection string and then it replaces with the same format?

**Comment by JohanPetersson** [comment]
> Do you mean I could name my file Test._Test.mdf instead and use the same format (Test._Test) in the connection string and then it replaces with the same format?

This did not work. So, I don't know how to interpret the tables here https://github.com/dotnet/templating/wiki/Naming-and-default-value-forms, other than that they describe which transformation will happen, but you cannot decide which one to use.

My file ended up with the name Test._Test.mdf and not new_name.mdf. Same thing in the json file.

Am I missing something?

**Reply by vlada-shubina** [reply]
`Test.Test` is still ambiguous name, because it is a valid namespace. In this case identity form will be "Test.Test" and namespace form will also be "Test.Test", that's why `Test._Test` was not replaced.

`Test.1` will not be ambiguous.
- identity form for this would be `Test.1`
- namespace form will be `Test._1`.

Could you please try that source name name and see if it works?

In you template content here you need to use identity form `Test.1` so that `-` is not replaced with `_`.
```
"ConnectionStrings": {
  "db": "Data Source=(LocalDb)\\MSSQLLocalDB;AttachDbFilename=|DataDirectory|\\Test.1.mdf;Initial Catalog=Test.Test;Integrated Security=True;Connect Timeout=30"
}
```

**Reply by JohanPetersson** [reply]
Thank you. That did the trick. Just so I understand, the magic here is '.1', right? The documentation says "like Template.1 does", but what else is unambiguous?

I've checked a lot of the templates you ship together with the SDK, none of them seems to use an unambiguous source name.

Not sure if it's just me, but the documentation around this is a bit confusing. Maybe it could be a bit more clear?

Noticed one thing, having leading underscore in namespaces renders warnings, see https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1707. 

Also naming project files e.g. Test.1.csproj looks a bit funky. But as long as it works...

**Reply by vlada-shubina** [reply]
Unambiguous source name should not be a valid namespace or class name, then the value forms for safe namespace and class name will replace invalid chars with underscores, and be different from identity value form. `Test.Test` is a valid namespace, so value form would be same as identity. In this case it is not 100% certain, which of these forms are applied when generating template content.

`Test.1` is not valid namespace and class name:
- namespace cannot have digit after `.`, so value form is Test._1`
- class name should not have `.` - value form is `Test__1`. 
- it also has small and capital letter to make a difference in lower case forms

I'm aware of SDK templates having ambiguous names, we already have some issues on that (https://github.com/dotnet/aspnetcore/issues/4081 as example). I created https://github.com/dotnet/templating/issues/4509 to improve the article. Imo, value forms are the most complicated topic in template authoring.

**Reply by JohanPetersson** [reply]
Would be easier to understand if the unambiguousness was achieved with a specific reserved word/token instead of digit. Something that is still valid namespace. Seems a bit weird that we should -- on purpose -- choose a source name that is not valid namespace.  

Again, thank you!
